posts/how-to-write-a-simproc.md

```diff
@@ -0,0 +1,370 @@
+But we are trying not to rely on the definition of `revRange`.
+
+**Pros**:
+* Doesn't require writing any meta code.
+* Doesn't require the recursion relations to be definitional (although they are in the case of `revRange`).
+
+**Cons**:
+* Requires adding two lemmas to your simp call instead of one (assuming we do not want these lemmas in the default simp set).
+* Simplifying `revRange n` for a big input numeral `n` might involve a lot of simplification steps.
+  In this specific case, the number of simplification steps is linear in `n`.
+  Simplification steps matter because each of them increases the size of the proof term.
+* `revRange n` could find itself (partially) evaluated even if `n` isn't a numeral.
+  Eg `simp [revRange_zero, revRange_succ]` on `⊢ revRange (n + 3) = revRange (3 + n)` will result in `⊢ n + 2 :: n + 1 :: n :: revRange n = revRange (3 + n)`.
+  This is in general highly undesirable.
+
+## The definitional approach
+
+In cases where the evaluation is definitionally equal to to the original expression, one may write a dsimproc instead of a simproc.
+The syntax to declare a dsimproc is rather to simprocs, with a small difference: we now need to return a `Simp.DStep` instead of a `Simp.Step`; in practice this amounts to providing the expression our program has produced without providing the proof (indeed, this is just `rfl`!)
+
+To compute `revRange` using the dsimproc approach, we can do the following:
+```lean
+dsimproc_decl revRangeCompute (revRange _) := fun e => do
+  --Extract the natural number from the expression
+  let_expr revRange m ← e | return .continue
+  --Recover the natural number as a term of type `Nat`
+  let some n ← Nat.fromExpr? m | return .continue
+  let l := revRange n
+  --Convert the list to an `Expr`
+  return .visit <| Lean.toExpr l
+```
+
+**Pros**:
+* Requires writing a single simproc.
+* Assuming the type of the expression to be evaluated implements `ToExpr`, there is no need to reevaluate the expression in the meta world.
+
+**Cons**:
+* The expression to be evaluated is traversed twice: Once to create its evaluation, then once more in the typechecking of the proof by `rfl`.
+* The produced `rfl` proof could be heavy.
+* Only works when the evaluation is definitionally equal to to the original expression.
+
+## The propositional approach
+
+A more general approach would be to manually construct the proof term we need to provide.
+In our case, we can do this in a recursive manner.
+```
+open Qq
+
+private theorem revRangeInduct {n : ℕ} {l : List ℕ}
+    (hl : revRange n = l) : revRange (n+1) = n :: l := by
+  induction n with
+  | zero => aesop
+  | succ n h => rw [←hl] ; rfl
+
+open Qq in
+simproc_decl revRangeComputeProp (revRange _) := fun e => do
+  let_expr revRange m ← e | return .continue
+  let some n ← Nat.fromExpr? m | return .continue
+  let rec go (n : ℕ) : (l : Q(List ℕ)) × Q(revRange $n = $l) :=
+    match n with
+    | 0 => ⟨q(([] : List ℕ)), q(rfl)⟩
```

Comment:
You are using a definitional equality here, that `revRange 0` is defined as the empty list!

Reply:
True! But only for the `n = 0` case - I figured never using definitional equalities would be a little cumbersome:)

Reply:
It's cumbersome but given that definitional equalities are explicitly mentioned plenty of times here, and the Pro you give is explicitly that you're not relying on definitional equalities, I think it's pretty confusing to keep that! The fix would be to prove `revRange_zero`, and use that as the proof instead of `rfl`.

Reply:
Doing that probably result in a proof that type-checks ever-so-slightly faster!